File: openhab/mounts/tmp/openhab/conf/automation/jsr223/python/personal/core_security.py
from __future__ import unicode_literals
from personal.core_helpers import reload_rules, get_date_string, get_all_semantic_items, get_semantic_items, intersection_count, sync_group_with_tags, get_date, has_same_location, METADATA_NAMESPACE
from core.triggers import when
from core.rules import rule
from personal.core_security import OperationState, is_security_state, ASSAULT_TRIGGER_EQUIPMENT_TAGS, ASSAULT_TRIGGER_POINT_TAGS, ASSAULT_DISARMER_EQUIPMENT_TAGS, ASSAULT_DISARMER_POINT_TAGS, LOCK_CLOSURE_EQUIPMENT_TAGS, LOCK_CLOSURE_POINT_TAGS, LOCK_EQUIPMENT_TAGS, LOCK_POINT_TAGS
from core.date import minutes_between, ZonedDateTime
from personal.core_broadcast import BroadcastType, broadcast
from core.jsr223.scope import ir, events, ON, OFF, OPEN, CLOSED
from org.openhab.core.types import UnDefType
from core.metadata import get_key_value, set_key_value
from org.openhab.core.model.script.actions import Log
from org.openhab.core.library.types import OnOffType, OpenClosedType


@rule("Core - Sync helper items of security", description="Core - Sync helper items", tags=['core', 'core-security'])
@when("Item added")
@when("Item updated")
@when("Item removed")
@when("System started")
def sync_security_helpers(event):
    # Sync group gCore_Security_AssaultTrigger with assault items - it's needed to create triggers on it
    sync_group_with_tags(
        ir.getItem("gCore_Security_AssaultTrigger"),
        ASSAULT_TRIGGER_EQUIPMENT_TAGS
    )

    # Sync group gCore_Security_AssaultDisarmamer with disarmer items - it's needed to create triggers on it
    sync_group_with_tags(
        ir.getItem("gCore_Security_AssaultDisarmamer"),
        ASSAULT_DISARMER_EQUIPMENT_TAGS
    )

    # Sync group gCore_Security_LockClosureTrigger with closure items - it's needed to create triggers on it
    sync_group_with_tags(
        ir.getItem("gCore_Security_LockClosureTrigger"),
        LOCK_CLOSURE_EQUIPMENT_TAGS
    )

    # Reload rules
    reload_rules(
        assault_trigger_triggers,
        assault_trigger
    )
    reload_rules(
        disarmament_triggers,
        disarmament
    )
    reload_rules(
        lock_closure_triggers,
        lock_closure
    )


assault_trigger_triggers = [
    "Descendent of gCore_Security_AssaultTrigger changed"
]


@rule("Core - Core_Security System - Trigger-Management", description="Core_Security System - Trigger-Management", tags=['core', 'core-security', 'core-reload-assault_trigger'])
@when("Descendent of gCore_Security_AssaultTrigger changed")
def assault_trigger(event):
    item = ir.getItem(event.itemName)
    if not is_security_state(OperationState.OFF) and (
        # Is target item:
        'gCore_Security_AssaultTrigger' in item.getGroupNames() or
        # Is Switch child of target item:
        intersection_count(item.getTags(), ASSAULT_TRIGGER_POINT_TAGS) > 0
    ):
        set_key_value(
            'Core_Security_OperationState',
            METADATA_NAMESPACE,
            'security',
            'last-alarm',
            get_date_string(ZonedDateTime.now())
        )

        message = "Silent alarm was triggered by {}!".format(item.label)
        if is_security_state(OperationState.ON):
            message = "Striking alarm was triggered by {}!".format(item.label)
            for alarm in ir.getItemsByTag("Alarm"):
                events.sendCommand(alarm, ON)

        broadcast(message, BroadcastType.ATTENTION)


@rule("Core - Core_Security System - Check Armament", description="Core_Security System - Check Armament", tags=['core', 'core-security'])
@when("Item Core_Security_OperationState received update {0}".format(OperationState.ON))
@when("Item Core_Security_OperationState received update {0}".format(OperationState.SILENTLY))
def armament(event):
    blockingAssaultTriggers = filter(
        lambda point: point.state in [OPEN, ON],
        reduce(
            lambda pointsList, newMember: pointsList + get_semantic_items(
                newMember,
                ASSAULT_TRIGGER_EQUIPMENT_TAGS,
                ASSAULT_TRIGGER_POINT_TAGS
            ),
            ir.getItem("gCore_Security_AssaultTrigger").members,
            []
        )
    )

    set_key_value(
        'Core_Security_OperationState',
        METADATA_NAMESPACE,
        'security',
        'blocking-assault-triggers',
        map(
            lambda trigger: trigger.name,
            blockingAssaultTriggers
        )
    )

    if blockingAssaultTriggers:
        events.postUpdate(
            ir.getItem('Core_Security_OperationState'),
            OperationState.OFF
        )
        broadcast("{} is in an OPEN state. No initiation into assault detection.".format(
            ", ".join(map(
                lambda trigger: trigger.label,
                blockingAssaultTriggers
            ))
        ))


disarmament_triggers = [
    "Descendent of gCore_Security_AssaultDisarmamer changed"
]


@rule("Core - Core_Security System - Disarmament-Management", description="Core_Security System - Disarmament-Management", tags=['core', 'core-security', 'core-reload-disarmament'])
@when("Descendent of gCore_Security_AssaultDisarmamer changed")
def disarmament(event):
    item = ir.getItem(event.itemName)
    if (
        # Is target item:
        'gCore_Security_AssaultDisarmamer' in item.getGroupNames() or
        # Is Switch child of target item:
        intersection_count(item.getTags(), ASSAULT_DISARMER_POINT_TAGS) > 0
    ) and (
        item.getStateAs(OnOffType) == ON or
        item.getStateAs(OpenClosedType) == OPEN
    ):
        events.postUpdate(
            ir.getItem("Core_Security_OperationState"),
            OperationState.OFF
        )


lock_closure_triggers = [
    "Descendent of gCore_Security_LockClosureTrigger changed"
]


@rule("Core - Core_Security System - Lock Closure-Management", description="Core_Security System - Lock Closure-Management", tags=['core', 'core-security', 'core-reload-lock_closure'])
@when("Descendent of gCore_Security_LockClosureTrigger changed")
def lock_closure(event):
    item = ir.getItem(event.itemName)
    if (
        # Is target item:
        'gCore_Security_LockClosureTrigger' in item.getGroupNames() or
        # Is Switch child of target item:
        intersection_count(item.getTags(), LOCK_CLOSURE_POINT_TAGS) > 0
    ) and (
        item.getStateAs(OnOffType) == OFF or
        item.getStateAs(OpenClosedType) == CLOSED
    ):
        for lock in get_all_semantic_items(LOCK_EQUIPMENT_TAGS, LOCK_POINT_TAGS):
            if has_same_location(item, lock):
                events.sendCommand(lock, ON)
                break


@rule("Core - Core_Security System - Turn off siren after Core_Security_OperationState update", description="Core_Security System - Turn off siren after Core_Security_OperationState update", tags=['core', 'core-security'])
@when("Item Core_Security_OperationState received update")
def siren_off(event):
    for alarm in ir.getItemsByTag("Alarm"):
        if alarm.getStateAs(OnOffType) != OFF:
            events.sendCommand(alarm, OFF)


@rule("Core - Core_Security System - Turn off siren after X minutes", description="Core_Security System - Turn off siren after X minutes", tags=['core', 'core-security'])
@when("Time cron 0 * * ? * * *")
def siren_autooff(event):
    autoOffTime = ir.getItem("Core_Security_SireneAutoOff")

    lastAlarmTime = get_key_value(
        'Core_Security_OperationState',
        METADATA_NAMESPACE,
        'security',
        'last-alarm'
    )

    if (
        isinstance(autoOffTime.state, UnDefType) or
        autoOffTime.state.floatValue() == 0 or
        not lastAlarmTime or
        minutes_between(
            get_date(lastAlarmTime),
            ZonedDateTime.now()
        ) > autoOffTime.state.floatValue()
    ):
        return

    for alarm in ir.getItemsByTag("Alarm"):
        if alarm.getStateAs(OnOffType) != OFF:
            events.sendCommand(alarm, OFF)
            broadcast(
                "Alarm item {} was automatically disabled after {} minutes.".format(
                    alarm.label,
                    autoOffTime.state
                ),
                BroadcastType.ATTENTION
            )

File: openhab/mounts/tmp/openhab/conf/automation/js/core-security.js
const { rules, items, triggers, time } = require('openhab')
const {
  json_storage,
  get_all_semantic_items,
  sync_group_with_semantic_items,
  broadcast,
  BroadcastType,
  stringifiedFloat,
  get_location,
  DATETIME_FORMAT,
  has_same_location
} = require(__dirname + '/core-helpers')

const OperationState = {
  OFF: 0,
  ON: 1,
  SILENTLY: 2
}

const ASSAULT_TRIGGER_EQUIPMENT_TAGS = ['Window', 'Door', 'CoreAssaultTrigger']
const ASSAULT_TRIGGER_POINT_TAGS = ['OpenState', 'Switch']
const ASSAULT_DISARMER_EQUIPMENT_TAGS = ['CoreAssaultDisarmer']
const ASSAULT_DISARMER_POINT_TAGS = ['OpenState', 'Switch']
const ASSAULT_ALARM_EQUIPMENT_TAGS = ['AlarmSystem', 'Siren']
const ASSAULT_ALARM_POINT_TAGS = ['Alarm']
const SMOKE_ALARM_EQUIPMENT_TAGS = ['SmokeDetector']
const SMOKE_ALARM_POINT_TAGS = ['Alarm']
const LOCK_CLOSURE_EQUIPMENT_TAGS = ['CoreLockClosure']
const LOCK_CLOSURE_POINT_TAGS = ['OpenState', 'Switch']
const LOCK_EQUIPMENT_TAGS = ['Lock']
const LOCK_POINT_TAGS = ['OpenState', 'Switch']

function is_security_state(state = OperationState.OFF) {
  const OperationStateItem = items.getItem('Core_Security_OperationState')
  return OperationStateItem.state == state
}

function scriptLoaded() {
  rules.JSRule({
    name: 'sync_security_helpers',
    description: 'Core (JS) - Sync helper items of security',
    tags: ['core', 'core-security'],
    triggers: [
      triggers.GenericCronTrigger('30 0/5 * ? * * *'),
      triggers.SystemStartlevelTrigger(100)
    ],
    execute: (event) => {
      // Sync group gCore_Security_AssaultTrigger with assault items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_AssaultTrigger',
        ASSAULT_TRIGGER_EQUIPMENT_TAGS,
        ASSAULT_TRIGGER_POINT_TAGS
      )

      // Sync group gCore_Security_SmokeTrigger with smoke detector items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_SmokeTrigger',
        SMOKE_ALARM_EQUIPMENT_TAGS,
        SMOKE_ALARM_POINT_TAGS
      )

      // Sync group gCore_Security_Locks with lock items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_Locks',
        LOCK_EQUIPMENT_TAGS,
        LOCK_POINT_TAGS
      )

      // Sync group gCore_Security_AssaultDisarmamer with disarmer items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_AssaultDisarmamer',
        ASSAULT_DISARMER_EQUIPMENT_TAGS,
        ASSAULT_DISARMER_POINT_TAGS
      )

      // Sync group gCore_Security_LockClosureTrigger with closure items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_LockClosureTrigger',
        LOCK_CLOSURE_EQUIPMENT_TAGS,
        LOCK_CLOSURE_POINT_TAGS
      )

      // Sync group gCore_Security_AssaultAlarms with assault alarm items - it's needed to create triggers on it
      sync_group_with_semantic_items(
        'gCore_Security_AssaultAlarms',
        ASSAULT_ALARM_EQUIPMENT_TAGS,
        ASSAULT_ALARM_POINT_TAGS
      )
    }
  })

  rules.JSRule({
    name: 'assault_detection',
    description: 'Core (JS) - Core_Security System - Assault Detection.',
    tags: ['core', 'core-security'],
    triggers: [
      triggers.GroupStateChangeTrigger('gCore_Security_AssaultTrigger')
    ],
    execute: (event) => {
      if (is_security_state(OperationState.OFF)) {
        return
      }

      const item = items.getItem(event.itemName)
      json_storage('Core_Security_OperationState').set(
        'security',
        'last-alarm',
        time.ZonedDateTime.now().format(DATETIME_FORMAT)
      )
      //TODO language
      let message = `Silent alarm was triggered by ${item.label}!`
      if (is_security_state(OperationState.ON)) {
        message = `Striking alarm was triggered by ${item.label}!`
        for (const alarm of get_all_semantic_items(
          ASSAULT_ALARM_EQUIPMENT_TAGS,
          ASSAULT_ALARM_POINT_TAGS
        )) {
          alarm.sendCommand('ON')
        }

        broadcast(message, BroadcastType.ATTENTION)
      }
    }
  })

  rules.JSRule({
    name: 'smoke_detection',
    description: 'Core (JS) - Core_Security System - Smoke Detection.',
    tags: ['core', 'core-security'],
    triggers: [triggers.GroupStateChangeTrigger('gCore_Security_SmokeTrigger')],
    execute: (event) => {
      const item = items.getItem(event.itemName)
      json_storage('Core_Security_OperationState').set(
        'security',
        'last-alarm',
        time.ZonedDateTime.now().format(DATETIME_FORMAT)
      )
      //TODO language
      const location = get_location(item)
      let message = `Smoke was detected ${
        location ? `in ${location.label}` : `by ${item.label}`
      }!`

      broadcast(message, BroadcastType.ATTENTION)
    }
  })

  rules.JSRule({
    name: 'armament',
    description: 'Core (JS) - Core_Security System - Check Armament.',
    tags: ['core', 'core-security'],
    triggers: [
      triggers.ItemStateUpdateTrigger(
        'Core_Security_OperationState',
        OperationState.ON
      ),
      triggers.ItemStateUpdateTrigger(
        'Core_Security_OperationState',
        OperationState.SILENTLY
      )
    ],
    execute: (event) => {
      const blockingAssaultTriggers = items
        .getItem('gCore_Security_AssaultTrigger')
        .members.reduce((pointsList, newMember) => {
          const newPoints = get_semantic_items(
            newMember,
            ASSAULT_TRIGGER_EQUIPMENT_TAGS,
            ASSAULT_TRIGGER_POINT_TAGS
          )
          return pointsList.concat(newPoints)
        }, [])
        .filter((point) => ['OPEN', 'ON'].some((state) => state == point.state))

      json_storage('Core_Security_OperationState').set(
        'security',
        'blocking-assault-triggers',
        blockingAssaultTriggers.map((trigger) => trigger.name).join(',')
      )
      //TODO language
      if (blockingAssaultTriggers.length > 0) {
        items
          .getItem('Core_Security_OperationState')
          .postUpdate(stringifiedFloat(OperationState.OFF))
        broadcast(
          `${', '.join(
            blockingAssaultTriggers.map((trigger) => trigger.label)
          )} is in an OPEN state. No initiation into assault detection.`
        )
      }
    }
  })

  rules.JSRule({
    name: 'disarmament',
    description: 'Core (JS) - Core_Security System - Disarmament-Management.',
    tags: ['core', 'core-security'],
    triggers: [
      triggers.GroupStateChangeTrigger('gCore_Security_AssaultDisarmamer')
    ],
    execute: (event) => {
      const item = items.getItem(event.itemName)
      if (['ON', 'OPEN'].some((state) => state == item.state)) {
        items
          .getItem('Core_Security_OperationState')
          .postUpdate(stringifiedFloat(OperationState.OFF))
      }
    }
  })

  rules.JSRule({
    name: 'lock_closure',
    description: 'Core (JS) - Core_Security System - Lock Closure-Management.',
    tags: ['core', 'core-security'],
    triggers: [
      triggers.GroupStateChangeTrigger('gCore_Security_LockClosureTrigger')
    ],
    execute: (event) => {
      const item = items.getItem(event.itemName)
      if (['OFF', 'CLOSED'].some((state) => state == item.state)) {
        for (const lock of get_all_semantic_items(
          LOCK_EQUIPMENT_TAGS,
          LOCK_POINT_TAGS
        )) {
          if (has_same_location(item, lock)) {
            lock.sendCommand('ON')
            break
          }
        }
      }
    }
  })

  rules.JSRule({
    name: 'siren_off',
    description:
      'Core (JS) - Turn off siren after Core_Security_OperationState update.',
    tags: ['core', 'core-security'],
    triggers: [triggers.ItemStateUpdateTrigger('Core_Security_OperationState')],
    execute: (event) => {
      for (const alarm of get_all_semantic_items(
        ASSAULT_ALARM_EQUIPMENT_TAGS,
        ASSAULT_ALARM_POINT_TAGS
      )) {
        if (alarm.state != 'OFF') {
          alarm.sendCommand('OFF')
        }
      }
    }
  })

  rules.JSRule({
    name: 'assault_alarm_autooff',
    description: 'Core (JS) - Turn off assault alarm after X minutes.',
    tags: ['core', 'core-security'],
    triggers: [triggers.GenericCronTrigger('0 * * ? * * *')],
    execute: (event) => {
      const autoOffTime = items.getItem('Core_Security_AssaultAlarmAutoOff')
      const lastAlarmTime = json_storage('Core_Security_OperationState').get(
        'security',
        'last-alarm'
      )

      if (
        autoOffTime.state == 0 ||
        !lastAlarmTime ||
        time.ZonedDateTime.parse(lastAlarmTime, DATETIME_FORMAT).until(
          time.ZonedDateTime.now(),
          time.ChronoUnit.MINUTES
        ) > autoOffTime.state
      ) {
        return
      }

      for (const alarm of get_all_semantic_items(
        ASSAULT_ALARM_EQUIPMENT_TAGS,
        ASSAULT_ALARM_POINT_TAGS
      )) {
        if (alarm.state != 'OFF') {
          alarm.sendCommand('OFF')
          //TODO language
          broadcast(
            `Alarm item ${alarm.label} was automatically disabled after ${autoOffTime.state} minutes.`,
            BroadcastType.ATTENTION
          )
        }
      }
    }
  })
}

module.exports = {
  is_security_state,
  OperationState,
  ASSAULT_TRIGGER_EQUIPMENT_TAGS,
  ASSAULT_TRIGGER_POINT_TAGS,
  ASSAULT_DISARMER_EQUIPMENT_TAGS,
  ASSAULT_DISARMER_POINT_TAGS,
  ASSAULT_ALARM_EQUIPMENT_TAGS,
  ASSAULT_ALARM_POINT_TAGS,
  LOCK_CLOSURE_EQUIPMENT_TAGS,
  LOCK_CLOSURE_POINT_TAGS,
  LOCK_EQUIPMENT_TAGS,
  LOCK_POINT_TAGS,
  SMOKE_ALARM_EQUIPMENT_TAGS,
  SMOKE_ALARM_POINT_TAGS
}
